ASoC: SOF: ipc4-topology: Advertise passthrough capable PCMs (using ChainDMA)

ujfalusi opened the pull request:
If a PCM is set to use ChainDMA then add it to the card->components string
    after a marker of iec61937-pcm:, for example on current HDA platforms where
    HDMI is set to use ChainDMA:
    iec61937-pcm:5,4,3 (the ordr of the PCM ids can differ)
    
    UCM is expected to parse and use this property to allow applications to
    use bytsestream passthrough in a standard way.

ujfalusi committed:
ASoC: SOF: ipc4-topology: Advertise passthrough capable PCMs (using ChainDMA)

If a PCM is set to use ChainDMA then add it to the card->components string
after a marker of iec61937-pcm:, for example on current HDA platforms where
HDMI is set to use ChainDMA:
iec61937-pcm:5,4,3 (the order of the PCM ids can differ)

UCM is expected to parse and use this property to allow applications to
use bytestream passthrough in a standard way.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

ujfalusi commented:
Changes since v1:
- replace 'hdmi-pt:1' with a simple 'hdmi-pt'

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -407,6 +407,39 @@ static void sof_ipc4_widget_update_kcontrol_module_id(struct snd_sof_widget *swi
+	/*
+	 * Update the card's components list with hdmi-pt (passthrough) if
+	 * ChainDMA is enabled for HDMI PCM stream
+	 */
```
well, technically the IEC61937 passthrough is not limited to HDMI. I use it with SPDIF on my UP2/Hifiberry setup to listen to Spotify. Some HDaudio platforms support it as well.

I think it should be a combination of use_chain_dma and back-end capabilities. Rejecting any PCM device that doesn't start with HDMI seems a bit strong. Or maybe we document the other options and explain why we are only doing this with HDMI for now.

ujfalusi replied:
> well, technically the IEC61937 passthrough is not limited to HDMI. I use it with SPDIF on my UP2/Hifiberry setup to listen to Spotify. Some HDaudio platforms support it as well.

I'm fairly certain that with up2+hifiberry you are not going to do a passthrough. It is converting the PCM from the DSP to S/PDIF output.

I'm also fairly certain that IPC4 is a rough ride on UP2

> I think it should be a combination of use_chain_dma and back-end capabilities. Rejecting any PCM device that doesn't start with HDMI seems a bit strong. Or maybe we document the other options and explain why we are only doing this with HDMI for now.

We don't really have setup other than HDMI/DP where we could use SOF+passthrough with SOF (IPC4).

plbossart replied:
I enabled passthrough over SPDIF in the past, that was actually my FIRST alsa patch back in 2010...
````
1965c441ecf1 ALSA: hda: enable SPDIF output for Conexant 5051/Lenovo docking stations
````

There is nothing specific about SPDIF or HDMI, it's all a regular PCM with additional bits in the right places.

You could absolutely test this feature on a device that has an SPDIF output.

So while I agree that we are limited in our testing to HDMI+IPC4 for now, the code is generic. We should at a minimum have a TODO or FIXME saying that the capability is restricted to HDMI.

The UCM part should absolutely not have any references to HDMI.

ujfalusi replied:
> I enabled passthrough over SPDIF in the past, that was actually my FIRST alsa patch back in 2010...
> 
> ```
> 1965c441ecf1 ALSA: hda: enable SPDIF output for Conexant 5051/Lenovo docking stations
> ```
> 
> There is nothing specific about SPDIF or HDMI, it's all a regular PCM with additional bits in the right places.
> 
> You could absolutely test this feature on a device that has an SPDIF output.

Sure, I'm using it on several devices around the house ;)

There is one big _if_ in here: only _if_ the firmware will not touch the data in any ways. It has to be bit accurate.
So far this is only guarantied with IPC4 and ChainDMA.
 
> So while I agree that we are limited in our testing to HDMI+IPC4 for now, the code is generic. We should at a minimum have a TODO or FIXME saying that the capability is restricted to HDMI.
> 
> The UCM part should absolutely not have any references to HDMI.

It has to have, user space expects the hdmi: mapping PCM for display, it might be OK to have iec958: for them, but we should keep that for non display digital outputs, like S/PDIF

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -407,6 +407,39 @@ static void sof_ipc4_widget_update_kcontrol_module_id(struct snd_sof_widget *swi
+	if (!strstr(spcm->pcm.pcm_name, "HDMI") || !pipeline->use_chain_dma)
+		return 0;
+
+	if (card->components && strstr(card->components, "hdmi-pt"))
```
maybe use iec61937=1

ujfalusi replied:
I don't think that is right, how does UCM will know which PCM we want to mark? What happens if we run on a device which do have real digital output other than the HDMI?
pcm3:pt pcm4:pt, etc
HDMI1:pt HDMI2:pt, etc
s/pt/iec985/

We could add string for each PCMs marked using ChainDMA, but not sure how that would be handled by UCM.
For HDMI passthrough we need `hdmi:`, for digital output we would need `iec958:` (not to mention those `surroundXY` maps).

ranj063 commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -407,6 +407,39 @@ static void sof_ipc4_widget_update_kcontrol_module_id(struct snd_sof_widget *swi
+
+	if (card->components)
+		card->components = devm_kasprintf(card->dev, GFP_KERNEL,
+						  "%s hdmi-pt",
```
@ujfalusi in the case that card->components is not NULL, dont you have to free it first before allocating a new one?

ujfalusi replied:
we never do that in sound/soc/intel/boards/ , but I agree.
It looks like we always use devm_kasprintf() for card->components, but if some code path will drop the devm_ then devm_kfree() will scream...
Let me add this to the patch.

ujfalusi replied:
We cannot free it as for example the skl_hda_dsp_generic.c is not allocating the initial string but assigns it to a static char* (hda_soc_components).

And we also have some funny issues with module remove/load.
This is with new version which records the PCM id which has ChainDMA enabled
```
# amixer -c0 info | grep Components
  Components    : 'HDA:80862812,80860101,00100000 HDA:10ec0888,160d0888,00100302 iec958-pcm:5,4,3'
# rmmod snd_sof_pci_intel_tgl; sleep 1; modprobe snd_sof_pci_intel_tgl
# amixer -c0 info | grep Components
  Components    : 'HDA:80862812,80860101,00100000 HDA:10ec0888,160d0888,00100302 iec958-pcm:5,4,3,5,4,3'
```
The module for the machine driver is not removed, so the `card->components` remains allocated and assigned, my code is then appends the PCM ids again when the topology is parsed for the second time, third time, etc.
snd-sdw is not doing this since it is brute force overwrites the `card->components` string on card creation with the "cfg-amp:X" write, it is just 'eating' up the memory silently.
It appears that removing `snd_soc_skl_hda_dsp` and probing it is not working well (at all), so, we have some fixes to do.

ujfalusi replied:
Hrm, actually it is worst....

ujfalusi replied:
@plbossart, @ranj063, @kv2019i: https://github.com/thesofproject/linux/pull/4928

ujfalusi commented:
Changes since v2:
- depends on https://github.com/thesofproject/linux/pull/4928 because now the old components string is freed.
- the new implementation will collect the PCM ids after a `iec958-pcm:` tag and user space is expected to use this to figure things out (on HDA board we know that PCM 3-5 is the HDMI for example)

ujfalusi commented:
Changes since v3:
- Updated the commit message to reflect the hdmi-pt -> iec958-pcm:<ids> change

plbossart commented on sound/soc/sof/ipc4-topology.c:
```diff
@@ -407,6 +407,52 @@ static void sof_ipc4_widget_update_kcontrol_module_id(struct snd_sof_widget *swi
 	}
 }
 
+static int
```
commit message typos
ordr
bytsestream

ujfalusi commented:
Changes since v5:
- fixed typos in commit message

ujfalusi commented:
@plbossart, @kv2019i, @ranj063, I guess the main question around this is: will it scale to S/PDIF PCM and how we will deal with compress? Compress is, well, compressed and it has it's own API to query what it supports, so... I think this should scale for normal PCMs.

ujfalusi commented:
Accompanying UCM patches: https://github.com/ujfalusi/alsa-ucm-conf/commits/topic/sof-hdmi/
